File: datastructures/stacksQueues/queue.js
import {Node} from "nodeClass.js"

class Queue {
    constructor() {
        this.first = null;
        this.last = null;
        this.length = 0;
    }

    peek() {
        return this.first;
    }

    enqueue(value) {
        const newNode = new Node(value);

        if (this.length === 0) {
            this.first = newNode;
            this.last = newNode;
        } else {
            this.last.next = newNode;
            this.last = newNode;
        }
        this.length++;
        return this;
    }

    dequeue() {
        if (this.length === 0) return null;
        if (this.length === 1) this.last = null;

        const first = this.first;
        this.first = this.first.next;

        this.length--;
        return first;
    }

    isEmpty() {
        if (this.length === 0) return true;
        return false;
    }
}

File: datastructures/stacksQueues/queue.py
from nodeClass import Node


class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0

    def __len__(self):
        return self.length

    def peek(self):
        return self.first

    def enqueue(self, value):
        NEW_NODE = Node(value)

        if len(self) == 0:
            self.first = NEW_NODE
            self.last = NEW_NODE
        else:
            self.last.next = NEW_NODE
            self.last = NEW_NODE
        
        self.length += 1
        return self

    def dequeue(self):
        if len(self) == 0:
            return None
        if len(self) == 1:
            self.last = None

        FIRST = self.first
        self.first = self.first.next

        self.length -= 1
        return FIRST

    def is_empty(self):
        if len(self) == 0:
            return True
        return False
